File: output_20260321_064953/qwen3/c/Struct/deposit.c
#include "stdlib.h"

/*@
predicate account(struct account *a; int balance) = a->balance |-> balance;
@*/

struct account *create_account()

{
    //@ requires true;
    //@ ensures account(result, 0);
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    myAccount->balance = 0;
    //@ close account(myAccount, 0);
    return myAccount;
}

void account_deposit(struct account *myAccount, int amount)

{
    //@ requires account(myAccount, ?b);
    //@ ensures account(myAccount, b + amount);
    //@ open account(myAccount, _);
    myAccount->balance += amount;
    //@ close account(myAccount, myAccount->balance);
}

int account_get_balance(struct account *myAccount)

{
    //@ requires account(myAccount, ?b);
    //@ ensures account(myAccount, b) &*& result == b;
    //@ open account(myAccount, _);
    int result = myAccount->balance;
    //@ close account(myAccount, result);
    return result;
}

void account_set_balance(struct account *myAccount, int newBalance)

{
    //@ requires account(myAccount, ?b);
    //@ ensures account(myAccount, newBalance);
    //@ open account(myAccount, _);
    myAccount->balance = newBalance;
    //@ close account(myAccount, newBalance);
}

void account_dispose(struct account *myAccount)

{
    //@ requires account(myAccount, _);
    //@ ensures true;
    //@ open account(myAccount, _);
    free(myAccount);
}

int main()

{
    //@ requires true;
    //@ ensures true;
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_deposit(myAccount, 10);
    int b = account_get_balance(myAccount);
    assert(b == 15);
    account_dispose(myAccount);
    return 0;
}

File: output_log/claude-opus/sample_3/rust/qwen3/rust/purely_unsafe/solutions/deposit.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

struct Account {
    balance: i32,
}

predicate acc(int *balance_ptr; int value) = balance_ptr |-> value;

predicate account_pred(Account *p) = 
    p != 0 &*& 
    malloc_block_Account(p, 1) &*& 
    p->balance |-> ?b &*& 
    acc(&(p->balance), b);

lemma void acc_read_write(int *balance_ptr, int old_value, int new_value)
    requires acc(balance_ptr, old_value);
    ensures acc(balance_ptr, new_value);
{
    open acc(balance_ptr, old_value);
    *balance_ptr = new_value;
    close acc(balance_ptr, new_value);
}

impl Account {
    unsafe fn create() -> *mut Account
    #[requires(true)]
    #[ensures(account_pred(result))]
    {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        (*my_account).balance = 0;
        close acc(&mut (*my_account).balance, 0);
        close account_pred(my_account);
        return my_account;
    }

    unsafe fn get_balance(my_account: *mut Account) -> i32
    #[requires(account_pred(my_account))]
    #[ensures(account_pred(my_account) &*& result == (*my_account).balance)]
    {
        open account_pred(my_account);
        open acc(&mut (*my_account).balance, _);
        int b = (*my_account).balance;
        close acc(&mut (*my_account).balance, b);
        close account_pred(my_account);
        return b;
    }

    unsafe fn set_balance(my_account: *mut Account, new_balance: i32)
    #[requires(account_pred(my_account))]
    #[ensures(account_pred(my_account) &*& (*my_account).balance == new_balance)]
    {
        open account_pred(my_account);
        open acc(&mut (*my_account).balance, _);
        (*my_account).balance = new_balance;
        close acc(&mut (*my_account).balance, new_balance);
        close account_pred(my_account);
    }

    unsafe fn deposit(my_account: *mut Account, amount: i32)
    #[requires(account_pred(my_account))]
    #[ensures(account_pred(my_account) &*& (*my_account).balance == old((*my_account).balance) + amount)]
    {
        open account_pred(my_account);
        open acc(&mut (*my_account).balance, ?old_balance);
        int new_balance = old_balance + amount;
        (*my_account).balance = new_balance;
        close acc(&mut (*my_account).balance, new_balance);
        close account_pred(my_account);
    }

    unsafe fn dispose(my_account: *mut Account)
    #[requires(account_pred(my_account))]
    #[ensures(true)]
    {
        open account_pred(my_account);
        open acc(&mut (*my_account).balance, _);
        close acc(&mut (*my_account).balance, _); // Close again for balance
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }
}

fn main() {
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::deposit(my_account, 10);
        let b = Account::get_balance(my_account);
        std::hint::assert_unchecked(b == 15);
        Account::dispose(my_account);
    }
}